apps/prairielearn/src/tests/groupExam.test.ts

```diff
@@ -0,0 +1,513 @@
+import { assert } from 'chai';
+import cheerio = require('cheerio');
+import fetch from 'node-fetch';
+import { config } from '../lib/config';
+import { step } from 'mocha-steps';
+
+import { queryAsync, queryOneRowAsync, loadSqlEquiv } from '@prairielearn/postgres';
+const sql = loadSqlEquiv(__filename);
+
+import helperServer = require('./helperServer');
+import { TEST_COURSE_PATH } from '../lib/paths';
+
+const siteUrl = 'http://localhost:' + config.serverPort;
+const baseUrl = siteUrl + '/pl';
+const courseInstanceUrl = baseUrl + '/course_instance/1';
+
+const storedConfig: any = {};
+
+const GROUP_NAME = 'groupBB';
+const GROUP_NAME_ALTERNATIVE = 'groupCC';
+
+interface StudentUser {
+  user_id: string;
+  uid: string;
+  name: string;
+  uin: string;
+}
+
+async function generateThreeStudentUsers(): Promise<StudentUser[]> {
+  const result = await queryAsync(sql.generate_and_enroll_3_users, []);
+  assert.lengthOf(result.rows, 3);
+  return result.rows;
+}
+
+/**
+ * Switches active user and loads assessment. Asserts the number of CSRF tokens on the page
+ * after loading the given assessment URL.
+ */
+async function switchUserAndLoadAssessment(
+  studentUser: StudentUser,
+  assessmentUrl: string,
+  authUin: string,
+  numCsrfTokens: number,
+): Promise<{ $: cheerio.Root; csrfToken: string }> {
+  // Load config
+  config.authUid = studentUser.uid;
+  config.authName = studentUser.name;
+  config.authUin = authUin;
+
+  // Load assessment
+  const res = await fetch(assessmentUrl);
+  assert.isOk(res.ok);
+  const page = await res.text();
+  const $ = cheerio.load(page);
+
+  // Check for CSRF tokens
+  const elemList = $('form input[name="__csrf_token"]');
+  assert.lengthOf(elemList, numCsrfTokens);
+  assert.nestedProperty(elemList[0], 'attribs.value');
+  assert.isString(elemList.attr('value'));
+  const csrfToken = elemList.attr('value') as string; // guaranteed to be string by assertion
+
+  return { $, csrfToken };
```

Comment:
It seems like this will find *any* CSRF token on the page, which doesn't test if the form actually being submitted has a CSRF token. Rather than this, it might be better to have the caller find the CSRF token in the form that the tests simulate a submission from:

```
const csrfTokenElement = $('form[name="name-of-form"] input[name="__csrf_token"]');
// ...
```

The `form[name="name-of-form"]` selector can be whatever is appropriate for the form in question.